File: 2018/python/solution3.py
"""
solution3.py
"""

import numpy as np

def readfile(fname='../data/input-3.txt'):
    with open(fname, 'r') as fin:
        return [l.strip() for l in fin]

def parseline(line):
    """
    Parse Elf area claim into coordinate ranges:
        #1 @ 509,796: 18x15
        -> [[509, 527], [796, 811]]
    """
    offset, widths = line.split('@')[1].strip().split(': ')
    offset = list(map(int, offset.split(',')))
    widths = list(map(int, widths.split('x')))
    return (slice(offset[1], offset[1]+widths[1]),
            slice(offset[0], offset[0]+widths[0]))

def parse_data(fname='../data/input-3.txt'):
    return [parseline(d) for d in readfile(fname)]

def count_claims(data):
    L = max(map(lambda x: max(x[0].stop, x[1].stop), data))
    claims = np.zeros((L+1, L+1))
    for d in data:
        claims[d] += 1
    return claims

def count_conflicts(claims):
    return np.sum(claims > 1)

def find_free_claim(data, claims):
    for i, d in enumerate(data):
        if np.all(claims[d] == 1):
            return i + 1

DATA = parse_data()
CLAIMS = count_claims(DATA)
print("Overlapping claims = {}".format(count_conflicts(CLAIMS)))
print("Free claim = {}".format(find_free_claim(DATA, CLAIMS)))

File: 2018/nim/solution3.nim
#[
#  Day 3
]#

import strutils, strformat, sequtils, sets

var seps: set[char] = {':', ',', 'x'}
seps.incl(Whitespace)

proc parseLine(line: string): seq[int]=
  var tokens = split(line, seps)
  var squares = @[parseInt(tokens[3]),  ## top offset
                  parseInt(tokens[6]),  ## height
                  parseInt(tokens[2]),  ## left offset
                  parseInt(tokens[5])]  ## width
  result.add(squares[0])
  result.add(squares[0] + squares[1])
  result.add(squares[2])
  result.add(squares[2] + squares[3])

proc parseFile(fname: string="../data/input-3.txt"): seq[seq[int]] =
  for line in open(fname).lines:
    result.add(parseLine(line.strip))

proc max(x: seq[seq[int]]): int=
  for y in x:
    result = max(result, max(y))

proc addClaim(claims: var seq[seq[int]], newclaim: seq[int]): int {.discardable.} =
  assert len(newclaim) == 4
  for row in newclaim[0]..<newclaim[1]:
    for col in newclaim[2]..<newclaim[3]:
      claims[row][col] += 1

proc countClaims(data: seq[seq[int]]): seq[seq[int]] =
  let L = max(data)
  result = newSeqWith(L, newSeq[int](L))
  for d in data:
    result.addClaim(d)

proc countConflicts(claims: seq[seq[int]]): int =
  let L = len(claims)
  for i in 0..<L:
    for j in 0..<L:
      if claims[i][j] > 1:
        inc(result)

proc findFreeClaim(data, claims: seq[seq[int]]): int =
  for i, d in data:
    var free = true
    for r in d[0]..<d[1]:
      for c in d[2]..<d[3]:
        if claims[r][c] > 1:
          free = false
    if free:
      return i + 1

let data = parseFile()
var claims = countClaims(data)
var conflicts = countConflicts(claims)
var freeclaim = findFreeClaim(data, claims)
echo fmt("Number conflicting claims = {conflicts}")
echo fmt("Free claim = {freeclaim}")
